pgjdbc/src/main/java/org/postgresql/jdbc/TypeInfoCache.java

```diff
@@ -395,11 +431,48 @@ public int getSQLType(int typeOid) throws SQLException {
       }
       rs.close();
 
+      if (sqlType == Types.STRUCT) {
+        loadStruct(typeOid);
+      }
       oidToSQLType.put(typeOid, sqlType);
       return sqlType;
     }
   }
 
+  private void loadStruct(int typeOid) throws SQLException {
+    PreparedStatement getCustomTypeFieldsStatement = prepareGetComplexTypeAttributesStatement();
+    getCustomTypeFieldsStatement.setLong(1, intOidToLong(typeOid));
+    if (!((BaseStatement) getCustomTypeFieldsStatement)
+        .executeWithFlags(QueryExecutor.QUERY_SUPPRESS_BEGIN)) {
+      throw new PSQLException(GT.tr("No results were returned by the query."), PSQLState.NO_DATA);
+    }
+
+    ResultSet rs = castNonNull(getCustomTypeFieldsStatement.getResultSet());
+
+    ArrayList<PgAttribute> pgAttributes = new ArrayList<>();
+    String typeName = null;
+    String arrayDelim = null;
+    while (rs.next()) {
+      if (typeName == null) {
+        typeName = rs.getString("type_name");
+        arrayDelim = rs.getString("array_delim");
+      }
+
+      String attrName = rs.getString("attr_name");
+      int attrOid = rs.getInt("attr_oid");
+      pgAttributes.add(new PgAttribute(attrName, castNonNull(getPGType(attrOid)), attrOid));
+    }
+    rs.close();
+
+    if (!pgAttributes.isEmpty()) {
+      PgStructDescriptor descriptor = new PgStructDescriptor(typeName, pgAttributes.toArray(new PgAttribute[0]));
+      pgOidToPgStructDescriptor.put(typeOid, descriptor);
+      pgNameToOid.put(typeName, typeOid);
+      arrayOidToDelimiter.put(typeOid, castNonNull(arrayDelim).charAt(0));
```

Comment:
Isnt there a different oid for the array than the base type (even for struct)?

Reply:
TypeInfoCache should be able to resolve both array type name and the array delimiter. I removed those lines as I think I added them when some tests were failing. It seems fine now.